api/hub/src/datasources/moderation-report.ts

```diff
@@ -0,0 +1,162 @@
+    const db: Client = await getAppDB();
+    const query = new Where('contentID')
+      .eq(contentID)
+      .skipNum(offset || 0)
+      .limitTo(limit || 10)
+      .orderByDesc('creationDate');
+    const reports = await db.find<ModerationReport>(this.dbID, this.collection, query);
+    const list = [];
+    for (const report of reports) {
+      report.explanation = decodeString(report.explanation);
+      list.push(report);
+    }
+    return list;
+  }
+
+  async getReport(contentID: string, author: string) {
+    const db: Client = await getAppDB();
+    const query = new Where('contentID').eq(contentID).and('author').eq(author);
+    const reports = await db.find<ModerationReport>(this.dbID, this.collection, query);
+    if (reports.length) {
+      reports[0].explanation = decodeString(reports[0].explanation);
+      return reports[0];
+    }
+    logger.warn(`report not found for contentID:${contentID} and author:${author}`);
+    return undefined;
+  }
+
+  async getFirstReport(contentID: string) {
+    const db: Client = await getAppDB();
+    // get the first report for this contentID
+    const query = new Where('contentID').eq(contentID).orderBy('creationDate').limitTo(1);
+    const reports = await db.find<ModerationReport>(this.dbID, this.collection, query);
+    if (reports.length) {
+      reports[0].explanation = decodeString(reports[0].explanation);
+      return reports[0];
+    }
+    logger.warn(`no reports found for contentID:${contentID}`);
+    throw new Error('report not found');
+  }
+
+  async getReasons(contentID: string) {
+    let reports: ModerationReport[];
+    const db: Client = await getAppDB();
+    const query = new Where('contentID').eq(contentID);
+    reports = await db.find<ModerationReport>(this.dbID, this.collection, query);
+    const reasons = [];
+    if (reports.length) {
+      for (const report of reports) {
+        // add unique reasons to the list
+        if (!reasons.includes(report.reason)) {
+          reasons.push(report.reason);
+        }
+      }
+    }
+    return reasons;
+  }
+
+  async addReport(
+    decisionsCollection: string,
+    contentType: string,
+    contentID: string,
+    author: string,
+    reason: string,
+    explanation?: string,
+  ) {
+    const db: Client = await getAppDB();
+    if (!author) {
+      throw new Error('Not authorized');
+    }
+
+    const exists = await this.exists(contentID, author);
+    if (exists) {
+      return Promise.reject(`cannot report content ${contentID} twice as user ${author}`);
+    }
+
+    // Do we need to check if a pending decision was created for this post?
+    const query = new Where('contentID').eq(contentID);
+    const results = await db.find<ModerationReport>(this.dbID, decisionsCollection, query);
+    if (results.length < 1) {
+      // need to create a pending decision to track moderation status
+      // how to avoid race condition here?
```

Comment:
you can have an `unique` index on `contentID` so there is just one decision record :)

Reply:
but I saw that this is already done so what's the remaining issue here?

Reply:
That is actually the case. I'm think there might be a possible race condition when two users report the same content and a pending decision is created. Unless `db.create` fails graciously in case a record exists.